docs/proposals/2841-flux-hpc/README.md

```diff
@@ -0,0 +1,185 @@
+    *   A **shared memory mount** that ensures the pod can utilize all shared memory on the node (or a reasonable subset). By default most container runtimes will mount only 64M, and this can negatively impact MPI performance.
+    *   **Shared `emptyDir` Volumes** are mounted into both the `initContainer` and the main application container to move the Flux install from the initContainer to the application container. The `initContainer` copies the Flux installation and necessary software from its own image into these shared volumes, and generates configuration for the cluster based on the user-preferences provided.
+    *   A **ConfigMap** is generated containing two scripts: `init.sh` (for the init container) and `entrypoint.sh` (for the main container). This ConfigMap is mounted into the pods.
+4.  **Execution Wrap**: The command of the user's main application container is overridden to use the `entrypoint.sh` script. This script first sets up the Flux environment (using the files from the shared volumes) and then uses `flux start` and `flux submit` to launch the user's original command within the now-active Flux cluster.
+5.  **Networking**: The plugin ensures the `JobSet` is configured with a headless service and DNS hostnames enabled, which Flux uses for its broker communication. High speed network for MPI can be used by way of extending pods to use a bypass mechanism to support Infiniband (Azure) or the Elastic Fabric Adapter (AWS).
+
+This approach provides an HPC environment without requiring the user to build a custom Docker image with Flux pre-installed, significantly improving the user experience.
+
+### API Design
+
+The proposed changes will integrate into the existing `v1alpha1` API structures. We will add a new field, `flux`, to the `MLPolicySource` struct. This aligns Flux with other runtimes like Torch and MPI.
+
+#### 1. `ClusterTrainingRuntime` and `TrainingRuntime`
+
+The proposal will leverage the existing `ClusterTrainingRuntime` (cluster-scoped) and `TrainingRuntime` (namespace-scoped) CRDs. These objects serve as reusable templates, and no changes are needed for their top-level structure.
+
+#### 2. Enhancing `MLPolicySource`
+
+We will add the `Flux` field to the `MLPolicySource` struct, which is embedded within the `MLPolicy`.
+
+```go
+type MLPolicySource struct {
+    Torch *TorchMLPolicySource `json:"torch,omitempty"`
+    MPI   *MPIMLPolicySource   `json:"mpi,omitempty"`
+
+    // FluxMLPolicy defines policy only for Flux
+    // +optional
+    Flux  *FluxMLPolicySource  `json:"flux,omitempty"`
+}
+```
+
+For the `FluxMLPolicySource`, we define the minimum required parameters needed for Flux and installing the view. The view must be compatible with the application container.
+
+```go
+// FluxMLPolicySource represents a Flux HPC runtime configuration.
+type FluxMLPolicySource struct {
```

Comment:
Do we also need: MPIImplementation  parameter in the  FluxMLPolicySource  API since we can support various MPI flavours like PMIx ?
cc @tenzen-y

Reply:
Generally no - most MPI implementations don't need customization to Flux flags. However, pmix does require a custom broker flag `-o pmi=pmix`. I would suggest we add that later when we consider more nuanced features.

Reply:
Sounds good!